### PR TITLE
Container build fixes + docker-compose

### PR DESCRIPTION
**Pull request by ebr**
Depends on #3547. Until that is merged, the container starts with  `ValueError: 'main' is not a valid ModelType` .

The PR can be tested locally by merging it with #3547 and running `docker-compose up --build` (see README for details)

Changes in this PR: 

- Replace Bash build scripts with `docker-compose`
- Rewrite Dockerfile:
  - enables running arbitrary commands (such as `invokeai-configure`) 
  - adds a build stage to build the UI, so that it's always up to date in the image
  - adds `magic-wormhole` and SSH support, useful for Runpod etc
  - no longer uses Docker volumes by default (user may still choose to do so via `docker-compose.yml`
  - an existing v3.0.0-compatible runtime directory can be mounted and used interchangeably with a native (non-docker) install
  - run as a non-privileged user; mounted runtime directory ownership will be retained
- Additional README and docs used in the Runpod template

Still TODO:
- update CI workflows for container building

### CODE REVIEW

File: docker/Dockerfile [hunk 1]
```diff
@@ -1,107 +1,122 @@
...
+
+FROM ubuntu:22.04 AS builder
+
+ARG DEBIAN_FRONTEND=noninteractive
+RUN rm -f /etc/apt/apt.conf.d/docker-clean; echo 'Binary::apt::APT::Keep-Downloaded-Packages "true";' > /etc/apt/apt.conf.d/keep-cache
+RUN --mount=type=cache,target=/var/cache/apt,sharing=locked \
+    --mount=type=cache,target=/var/lib/apt,sharing=locked \
+    apt update && apt-get install -y \
+        git \
+        python3-venv \
```
Comment (brandonrising) [inline docker/Dockerfile:13]: How are we ensuring it doesn't install python 3.11+?

Comment (ebr) [reply docker/Dockerfile:13]: `python3-venv` will match the installed python version. we can pin the version to be sure but i don't think it's necessary

**Comment by briancaffey** [comment]
I tried following the instructions in `docker/README.md` (Ubuntu 22.04 with NVIDIA GPU) and I'm getting the following error: 

```
 => [invokeai] exporting to image                                                                                                13.6s
 => => exporting layers                                                                                                          13.6s
 => => writing image sha256:cb67d1b9892a852ec86b6e341ad17c1e8b6301f255e285ef851734418c3de3fd                                      0.0s
 => => naming to docker.io/local/invokeai:latest                                                                                  0.0s
[+] Running 2/2
 ✔ Network docker_default       Created                                                                                           0.1s 
 ✔ Container docker-invokeai-1  Created                                                                                           0.1s 
Attaching to docker-invokeai-1
docker-invokeai-1  | Loading Python libraries...
docker-invokeai-1  | 
docker-invokeai-1  | ** INITIALIZING INVOKEAI RUNTIME DIRECTORY **
docker-invokeai-1  | 
docker-invokeai-1  | ** CHECKING/UPDATING SUPPORT MODELS **
docker-invokeai-1  | Installing bert tokenizer...
docker-invokeai-1  | ╭───────────────────── Traceback (most recent call last) ──────────────────────╮
docker-invokeai-1  | │ /opt/venv/invokeai/bin/invokeai-configure:8 in <module>                      │
docker-invokeai-1  | │                                                                              │
docker-invokeai-1  | │   5 from invokeai.frontend.install import invokeai_configure                 │
docker-invokeai-1  | │   6 if __name__ == '__main__':                                               │
docker-invokeai-1  | │   7 │   sys.argv[0] = re.sub(r'(-script\.pyw|\.exe)?$', '', sys.argv[0])     │
docker-invokeai-1  | │ ❱ 8 │   sys.exit(invokeai_configure())                                       │
docker-invokeai-1  | │   9                                                                          │
docker-invokeai-1  | │                                                                              │
docker-invokeai-1  | │ /opt/invokeai/invokeai/backend/install/invokeai_configure.py:867 in main     │
docker-invokeai-1  | │                                                                              │
docker-invokeai-1  | │   864 │   │   │   print("\n** SKIPPING SUPPORT MODEL DOWNLOADS PER USER REQU │
docker-invokeai-1  | │   865 │   │   else:                                                          │
docker-invokeai-1  | │   866 │   │   │   print("\n** CHECKING/UPDATING SUPPORT MODELS **")          │
docker-invokeai-1  | │ ❱ 867 │   │   │   download_bert()                                            │
docker-invokeai-1  | │   868 │   │   │   download_sd1_clip()                                        │
docker-invokeai-1  | │   869 │   │   │   download_sd2_clip()                                        │
docker-invokeai-1  | │   870 │   │   │   download_realesrgan()                                      │
docker-invokeai-1  | │                                                                              │
docker-invokeai-1  | │ /opt/invokeai/invokeai/backend/install/invokeai_configure.py:188 in          │
docker-invokeai-1  | │ download_bert                                                                │
docker-invokeai-1  | │                                                                              │
docker-invokeai-1  | │   185 │   │   warnings.filterwarnings("ignore", category=DeprecationWarning) │
docker-invokeai-1  | │   186 │   │   from transformers import BertTokenizerFast                     │
docker-invokeai-1  | │   187 │   │                                                                  │
docker-invokeai-1  | │ ❱ 188 │   │   download_from_hf(BertTokenizerFast, "bert-base-uncased")       │
docker-invokeai-1  | │   189                                                                        │
docker-invokeai-1  | │   190                                                                        │
docker-invokeai-1  | │   191 # ---------------------------------------------                        │
docker-invokeai-1  | │                                                                              │
docker-invokeai-1  | │ /opt/invokeai/invokeai/backend/install/model_install_backend.py:288 in       │
docker-invokeai-1  | │ download_from_hf                                                             │
docker-invokeai-1  | │                                                                              │
docker-invokeai-1  | │   285 │   logger = InvokeAILogger.getLogger('InvokeAI')                      │
docker-invokeai-1  | │   286 │   logger.addFilter(lambda x: 'fp16 is not a valid' not in x.getMessa │
docker-invokeai-1  | │   287 │                                                                      │
docker-invokeai-1  | │ ❱ 288 │   path = config.cache_dir                                            │
docker-invokeai-1  | │   289 │   model = model_class.from_pretrained(                               │
docker-invokeai-1  | │   290 │   │   model_name,                                                    │
docker-invokeai-1  | │   291 │   │   cache_dir=path,                                                │
docker-invokeai-1  | ╰──────────────────────────────────────────────────────────────────────────────╯
docker-invokeai-1  | AttributeError: 'InvokeAIAppConfig' object has no attribute 'cache_dir'
docker-invokeai-1 exited with code 1
```

I couldn't find any related issues to this error message. I'm starting with an empty `INVOKEAI_ROOT`. I think I've seen this error before on the `main` branch as well, but I'm not sure how to resolve it. Do you have any suggestions @ebr? Thanks!

**Comment by ebr** [comment]
> `docker-invokeai-1  | Exception: Unknown model type: main`

@briancaffey the branch was out of date with `main` - i just updated it, and it should fix the above issue. Yes, you can install some models now using the new `invokeai-model-install` script, and should be able to use this runtime dir with both containerized and bare-metal Invoke.

**Comment by briancaffey** [comment]
Thanks @ebr, I had pulled in `main` but did not rebuild the container. I was able to install models by setting this command in the docker-compose file: 

```
    command:
      - bash
      - -c
      - |
        invokeai-configure
```

and then running `docker compose run -it invokeai`. I wasn't able to use arrow keys to navigate the `invokeai-configure` menu when running this directly with `docker compose up`. 
 
One small suggestion would be to use `docker compose` instead of `docker-compose` in the docs and scripts. I'm excited to try out the new nodes interface. I'm also really interested in working on a Pulumi/Terraform component for deploying InvokeAI on AWS with ECS, Fargate and ECS.

Thank you!

File: docker/README.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+1. Ensure Docker has at least 16GB RAM
+2. Enable VirtioFS for file sharing
+3. Enable `docker-compose` V2 support
+
+This is done via Docker Desktop preferences
+
+## Quickstart
+
+
+1. Make a copy of `env.sample` and name it `.env` (`cp env.sample .env` (Mac/Linux) or `copy example.env .env` (Windows)). Make changes as necessary. Set `INVOKEAI_ROOT` to an absolute path to:
```
Comment (ElRoberto538) [inline docker/README.md:22]: The env.sample file doesn't seem to have been committed?

Comment (ElRoberto538) [reply docker/README.md:22]: Oops, it's just a typo...
```suggestion
1. Make a copy of `.env.sample` and name it `.env` (`cp .env.sample .env` (Mac/Linux) or `copy .env.sample .env` (Windows)). Make changes as necessary. Set `INVOKEAI_ROOT` to an absolute path to:
```

**Comment by fat-tire** [comment]
Trying w/ podman-compose (`apt get podman-compose`), got this:
```
[2/3] STEP 4/5: RUN --mount=type=cache,target=node_modules     npm install --include dev
Error: building at STEP "RUN --mount=type=cache,target=node_modules npm install --include dev": resolving mountpoints for container "[hash]": invalid container path "node_modules", must be an absolute path
exit code: 125
```
Dunno if this is a podman thing or not, but this seemed to fix it:
```diff
-RUN --mount=type=cache,target=node_modules \
+RUN --mount=type=cache,target=/usr/lib/node_modules \
     npm install --include dev
-RUN --mount=type=cache,target=node_modules \
+RUN --mount=type=cache,target=/usr/lib/node_modules \
```

Then ([as before with podman](https://github.com/cockpit-project/cockpit-podman/issues/166)) got:
```
[3/3] STEP 11/21: COPY --link --from=builder ${INVOKEAI_SRC} ${INVOKEAI_SRC}
Error: building at STEP "COPY --link --from=builder ${INVOKEAI_SRC} ${INVOKEAI_SRC}": COPY only supports the --chmod=<permissions> --chown=<uid:gid> and the --from=<image|stage> flags
exit code: 125
```
So had to remove the `--link`s.  (Wonder if this could just check if it's supported before using it and fallback gracefully?)

After slightly modifying my `run.sh` to remap to new volumes (the old ones wouldn't work as they were using python 3.9, getting rid of `--user=invokeai:invokeai`, setting  the caches as they have changed significantly (difference between `/opt/invokeai` and `/invokeai vs `/workspace` and the multiple `.cache` directories were a bit confusing?), and setting local user permissions as per the previous docker changes for podman), it came up.

Unfortunately, it got a "server error "trying to render anything.  So I also needed to manually install the nvidia-drivers as before.

Once this settles down and nightlys go out, I'll probably just base off the nightlies, then add the nvidia-drivers myself, but wanted to share this with anyone who might care.

PS-- Dang, nodes are glorious. [All I hoped and dreamed](https://github.com/invoke-ai/InvokeAI/discussions/597#discussioncomment-3658538), but better!  GREAT JOB everyone!

(Now if I can only figure out what to plug into what. Feel like I'm almost there, but still getting a "server error" on invoke.  This is definitely super-cool though.)

**Comment by ebr** [comment]
@fat-tire @ElRoberto538 @briancaffey thank you for your feedback thus far! i've incorporated most of your suggestions and this should be ready to merge.

@fat-tire: I tried to base the image on the `nvidia-cuda` to save the trouble of installing nvidia drivers, but that bloats the image to 11GB, so I rolled that back. And we'd like to keep `COPY --link` because it makes layer sharing between stages much more efficient. I recognize this is a blocker for `podman`, but perhaps you'd like contribute a working Dockerfile and a `podman-compose` setup in a separate PR once this is merged? We don't want to sacrifice the `buildkit` optimizations, but we don't want to leave `podman` users without support either. So please feel free to PR your improvements.

**Comment by fat-tire** [comment]
> @fat-tire @ElRoberto538 @briancaffey thank you for your feedback thus far! i've incorporated most of your suggestions and this should be ready to merge.

Awesome!

> @fat-tire: I tried to base the image on the `nvidia-cuda` to save the trouble of installing nvidia drivers, but that bloats the image to 11GB, so I rolled that back.

No problem... Once you set up the daily builds again, I will probably just continue to start with the "official" builds and then use a podman-specific `Dockerfile` that just uses the official build as a base and installs the drivers on top.  I had to modify the `run.sh` from what I was previously using to account for the new mountpoints and username changes and such, but it overall seems to work fine.

> And we'd like to keep `COPY --link` because it makes layer sharing between stages much more efficient. I recognize this is a blocker for `podman`, but perhaps you'd like contribute a working Dockerfile and a `podman-compose` setup in a separate PR once this is merged? We don't want to sacrifice the `buildkit` optimizations, but we don't want to leave `podman` users without support either. So please feel free to PR your improvements.

Once everything is settled maybe I can just add a Podman `Dockerfile` on top of what's there.  The current `Dockerfile` wouldnl't be usable because of the `--link`, but that image could be used as a base...

If I'm the only one out there using Podman (?), then it's not needed at all-- I just mostly wanted to document the Podman issues in case anyone else out there was trying it.

Thx!

**Pull request merged**